lib/rbs/type_alias_dependency.rb

```diff
@@ -0,0 +1,105 @@
+module RBS
+  class TypeAliasDependency
+    attr_reader :env
+
+    # Direct dependencies corresponds to a directed graph
+    # with vertices as types and directions based on assignment of types
+    attr_reader :direct_dependencies
+    # A hash which stores the transitive closure
+    # of the directed graph
+    attr_reader :dependencies
+
+    def initialize(env:)
+      @env = env
+    end
+
+    # Check if an alias type definition is circular & prohibited
+    def circular_definition?(alias_name)
+      # Construct transitive closure, if not constructed already
+      transitive_closure() unless @dependencies
+
+      # Check for recursive type alias
+      @dependencies[alias_name][alias_name]
+    end
+
+    def build_dependencies
+      return if @direct_dependencies
+
+      # Initialize hash(a directed graph)
+      @direct_dependencies = {}
+      # Initialize dependencies as an empty hash
+      @dependencies = {}
+      # Iterate over alias declarations inserted into environment
+      env.alias_decls.each do |name, entry|
+        # Construct a directed graph by recursively extracting type aliases
+        @direct_dependencies[name] = direct_dependency(entry.decl.type)
+        # Initialize dependencies with an empty hash
+        @dependencies[name] = {}
+      end
+    end
+
+    def transitive_closure
+      # Construct a graph of direct dependencies
+      build_dependencies()
+      # Construct transitive closure by using DFS(recursive technique)
+      @direct_dependencies.each_key do |name|
+        dependency(name, name)
+      end
+    end
+
+    def strongly_connected_components
+      # Contrsuct a graph of direct dependencies
+      build_dependencies()
+      # Create a proc for iterating over nodes
+      each_node = -> (&block) { @direct_dependencies.each_key(&block) }
+      # Create a proc for iterating over children of a node
+      each_child = -> (name, &block) { @direct_dependencies[name]&.each(&block) }
+      # Build a hash table of type names & strongly connected aliases
+      scc_hash = {}
+      # Apply TSort technique & fetch strongly connected components
+      strongly_connected_components = TSort.strongly_connected_components(each_node, each_child)
+      strongly_connected_components.each do |scc|
+        if scc.size > 1
+          scc_hash[name] = scc
+        end
+      end
+
+      scc_hash
+    end
+
+    private
+
+    # Constructs directed graph recursively
+    def direct_dependency(type, result = Set[])
```

Comment:
I'm not sure if this definition of `#direct_dependency` allows type definitions like:

```rbs
type a = Array[a]
type b = [b, b]
type c = { foo: c }
type d = ^(d) -> d
```

`#direct_dependency` will test if the type is a union, optional, or intersection and then returns the `#direct_dependency`s of the direct child types.